src/dso/dso_transport.hpp

```diff
@@ -0,0 +1,251 @@
+/*
+ *    Copyright (c) 2023, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for DSO agent.
+ */
+
+#ifndef OTBR_AGENT_DSO_TRANSPORT_HPP_
+#define OTBR_AGENT_DSO_TRANSPORT_HPP_
+
+#if OTBR_ENABLE_DNS_DSO
+
+#include <arpa/inet.h>
+#include <map>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+#include <cassert>
+#include <list>
+
+#include "common/code_utils.hpp"
+#include "common/mainloop.hpp"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/debug.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/error.h"
+#include "mbedtls/net_sockets.h"
+#include "mbedtls/ssl.h"
+#include "ncp/ncp_openthread.hpp"
+#include "openthread/logging.h"
+#include "openthread/message.h"
+#include "openthread/platform/dso_transport.h"
+
+namespace otbr {
+namespace Dso {
+
+// TODO: Support DNS-over-TLS
+/**
+ * @addtogroup border-router-dso
+ *
+ * @brief
+ *   This module includes definitions for DSO agent.
+ *
+ * @{
+ */
+
+class DsoAgent : public MainloopProcessor
```

Comment:
Consider writing unittests to ensure the quality? For example, make sure we handle errors properly? We can start two DsoAgent and inject errors in the tests.

Reply:
Added some unit test cases.